### PR TITLE
Adding BDP and Fixing Bugs

### PR DESCRIPTION
**Pull request by ASzot**

### CODE REVIEW

File: habitat-baselines/habitat_baselines/config/experiments_hab3/bdp_kinematic_oracle.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# @package _global_
```
Comment (jimmytyyang) [inline habitat-baselines/habitat_baselines/config/experiments_hab3/bdp_kinematic_oracle.yaml:1]: HI Andrew! thank you for working on this. What is "bdp" in the title? maybe we want to add one linear to talk about this in this yaml.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 2]
```diff
@@ -361,6 +361,7 @@ class AgentAccessMgrConfig(HabitatBaselinesBaseConfig):
     num_active_agents_per_type: List[int] = field(default_factory=lambda: [1])
     num_pool_agents_per_type: List[int] = field(default_factory=lambda: [1])
     agent_sample_interval: int = 20
+    behavior_latent_dim: int = -1
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:364]: What does a value of -1 mean herE?

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ROBOT_TYPE = 0
+HUMAN_TYPE = 1
+
+BEHAV_ID = "behav_latent"
+
+
+@baseline_registry.register_agent_access_mgr
+class BdpAgentAccessMgr(MultiAgentAccessMgr):
+    def _sample_active_idxs(self):
+        assert not self._pop_config.self_play_batched
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:47]: Can you add a comment here explaining that for BDP, the behav agent will always have the same policy?

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self._behav_latents = torch.zeros(
+            (num_envs, self._pop_config.behavior_latent_dim), device=device
+        )
+        return np.array([COORD_AGENT, BEHAV_AGENT]), np.array(
+            [ROBOT_TYPE, HUMAN_TYPE]
+        )
+
+    def _inject_behav_latent(self, obs, agent_idx):
+        agent_obs = update_dict_with_agent_prefix(obs, agent_idx)
+        if agent_idx == BEHAV_AGENT:
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:62]: will it give problems that all the observations start with `agent_X` but the behavior obs does not ?

Comment (xavierpuigf) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:62]: I think you should add an agent prefic to behav_id as well?

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py [hunk 5]
```diff
@@ -120,14 +127,24 @@ def get_extra(
...
+    ):
+        return MultiPolicy(update_obs_with_agent_prefix_fn)
 
 
 class MultiStorage(Storage):
-    def __init__(self):
+    def __init__(self, update_obs_with_agent_prefix_fn):
         self._active_storages = []
         self._agent_type_ids = []
+        if update_obs_with_agent_prefix_fn is None:
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:145]: Why do you need this now? Before we always had the agent prefix and didn't define this function. What changed?

**Review by xavierpuigf — COMMENTED** [review]
See my comments. Biggest comment is that I don't see anywhere in the policy that you condition on the agent embedding, even though pass that around. I also don't see where you do the sampling of the agent embedding.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py (hunk 2 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/default_structured_configs.py:372]: Added comment.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py (hunk 4 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:76]: No because this is the post-filtered observation dictionary with the `agent_` prefix already removed.

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py (hunk 5 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:145]: I need this so I can inject more information into the observation.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py (hunk 3 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:58]: Added doc string for this.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 6]
```diff
@@ -319,6 +319,13 @@ class AuxLossConfig(HabitatBaselinesBaseConfig):
...
 
+@dataclass
+class BdpDiscrimConfig(AuxLossConfig):
+    loss_scale: float = 0.1
+    hidden_size: int = 128
+    behavior_latent_dim: int = -1
+
+
 @dataclass
 class CPCALossConfig(AuxLossConfig):
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:330]: Isn't this name misleading? If I understand the code, you are currently using a classification loss, not CPC.

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 7]
```diff
@@ -71,29 +71,43 @@ def __init__(
...
-            self._active_agents = [0, 0]
+            self._active_agents = np.array([0, 0])
         else:
             policy_cls = MultiPolicy
             updater_cls = MultiUpdater
             storage_cls = MultiStorage
 
         # TODO(xavi to andrew): why do we call these functions? It seems they
         # just create an empty class
-        num_active_agents = sum(self._pop_config.num_active_agents_per_type)
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:86]: Same comment as before, it seems like these classes do not initialize the arguments you are passing.

**Review by xavierpuigf — COMMENTED** [review]
Generally looks good! What I don't understand is that the agent policy is not conditioned on the agent type anywhere, so there is no way the policy will differ for different agent types.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+            nn.ReLU(True),
+            nn.Linear(hidden_size, behavior_latent_dim),
+        )
+        self.loss_scale = loss_scale
+
+    def pred_logits(self, policy_features, obs):
+        return self.discrim(policy_features)
+
+    def forward(self, policy_features, obs):
+        pred_logits = self.pred_logits(policy_features, obs)
```
Comment (akshararai) [inline habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:178]: can't pass obs, since it contains the behavior latent. Remove behavior latent before passing to pred_logits.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py (hunk 6 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/default_structured_configs.py:330]: What do you mean by CPC? I shouldn't have added anything related to CPC?

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py (hunk 5 above)
Comment (akshararai) [reply habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:145]: Extra information is the behavior latent

**Review by akshararai — APPROVED** [review]
Updated some comments, and some possible changes. but looks good to me overall. Thanks for fixing all this!

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py (hunk 6 above)
Comment (xavierpuigf) [reply habitat-baselines/habitat_baselines/config/default_structured_configs.py:330]: Sorry, I thought the CPCALoss had been added.

**Review by xavierpuigf — APPROVED** [review]

File: habitat-baselines/habitat_baselines/rl/hrl/hl/neural_policy.py [hunk 9]
```diff
@@ -81,14 +100,21 @@ def __init__(self, *args, **kwargs):
         self._policy = CategoricalNet(self._hidden_size, self._n_actions)
         self._critic = CriticHead(self._hidden_size)
 
+        self._aux_modules = get_aux_modules(
+            aux_loss_config, action_space, self
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/hrl/hl/neural_policy.py:104]: We actually need to pass all of self so we can compute the hidden features.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        use_resume_state,
+        num_envs,
+        percent_done_fn,
+        lr_schedule_fn,
+        agent_name,
+    ):
+        if agent_name == BEHAV_AGENT_NAME:
+            # Inject the behavior latent into the observation spec
+            agent_env_spec.observation_space[BEHAV_ID] = spaces.Box(
+                low=np.finfo(np.float32).min,
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:119]: Ideally yes, but I am not sure how well Obs space int type support works in habitat.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        net: Net,
+        loss_scale,
+        hidden_size,
+        behavior_latent_dim,
+    ):
+        super().__init__()
+
+        # For now the discriminator input is just (z_t, a_t) where z_t is the
+        # LSTM encoded state and a_t is the current action.
+        input_dim = net._hidden_size
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:166]: Comment is inaccurate for now. Updated.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py (hunk 8 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:180]: We actually need to keep it if we decide to generate policy actions in the future (since the policy is latent conditioned).

File: habitat-baselines/habitat_baselines/rl/ppo/updater.py [hunk 12]
```diff
@@ -14,3 +18,17 @@ def update(self, rollouts: Storage) -> Dict[str, float]:
         """
         Perform an update from data in the storage objet.
         """
+
+    def init_distributed(self, find_unused_params: bool = True) -> None:
+        """
+        Setup distributed training.
+        """
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/ppo/updater.py:25]: You're right, my bad. Should be abstract.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            features, _ = self._hl_policy(
+                obs.map(lambda x: x.flatten(0, 1)),
+                behav_storage.buffers["recurrent_hidden_states"].flatten(0, 1),
+                masks.flatten(0, 1),
+            )
+            features = features.view(*masks.shape[:2], -1)
+            pred_logits = self._discrim.pred_logits(features, obs)
+            behav_ids = torch.argmax(obs[BEHAV_ID], -1).long()
+            scores = F.log_softmax(pred_logits, -1)
+            log_prob = scores.gather(-1, behav_ids.view(*masks.shape[:2], 1))
```
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:220]: One is for calculating the discriminator loss and the other is for calculating the diversity reward.

File: habitat-baselines/habitat_baselines/rl/ppo/updater.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/ppo/updater.py:26]: Actually I take this back. This is intentional to have no implementation. It should be empty.

File: habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py (hunk 8 above)
Comment (akshararai) [reply habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py:181]: The discriminator cannot have the latent variable as input. Otherwise it will ignore all inputs and just use the latent variable to discriminate the different policies? Am I misunderstanding this somehow?
